RavenDB-8113 Counters

aviv86 opened the pull request:
fix PR comments :
- change tryMergeAttachmentsConflict -> tryMergeMetadataConflicts
- fix bug in IsMetadataEqualTo()
- add HasCounters flag
- move RecreateCounters() to PUT
- save/delete counters from storage before updating document metadata

ayende commented on src/Raven.Server/Documents/DocumentCompare.cs:
```diff
@@ -56,30 +56,31 @@ private static DocumentCompareResult IsMetadataEqualTo(BlittableJsonReaderObject
-                    // If the conflict is just on @metadata with @attachment we know how to resolve it.
-                    if (currentMetadata.Count == 1 && currentMetadata.GetPropertyNames()[0].Equals(Constants.Documents.Metadata.Attachments, StringComparison.OrdinalIgnoreCase))
+                    // If there is a conflict on @metadata with @counters and/or with @attachments, we know how to resolve it.
+                    if (currentMetadata.GetPropertyNames().Contains(Constants.Documents.Metadata.Counters, StringComparer.OrdinalIgnoreCase))
```
Only call `currentMetadata.GetPropertyNames()` once

ayende commented on src/Raven.Server/Documents/DocumentPutAction.cs:
```diff
@@ -402,6 +422,91 @@ private static void ThrowInvalidDocumentId(string id)
+
+            var countersStorage = context.DocumentDatabase.DocumentsStorage.CountersStorage;
+            var onDiskCounters = countersStorage.GetCountersForDocument(context, id, 0, int.MaxValue);
+            var sortedCounters = new HashSet<string>(onDiskCounters, StringComparer.OrdinalIgnoreCase).ToList();
```
Don't need the `HashSet` here. 
Also, pretty sure that we already get them sorted at this point.

aviv86 committed:
RavenDB-8113 Counters - fix PR comments :
- change tryMergeAttachmentsConflict -> tryMergeMetadataConflicts
- fix bug in IsMetadataEqualTo()
- add HasCounters flag
- move RecreateCounters() to PUT
- save/delete counters from storage before updating document metadata
